### PR TITLE
[MRG] Supervised Learning benchmark

### PR DESCRIPTION
**Pull request by iaroslav-ai**
This is a continuation of https://github.com/scikit-optimize/scikit-optimize/pull/317 .
@MechCoder It seemed like the easiest solution was to just make a new fork of the repo and to copy the bench_ml.py file there. Otherwise I risk messing up things again with commits :)

### CODE REVIEW

File: benchmarks/bench_ml.py [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+
+
+    def evaluate(self, point):
+        """
+        Evaluates model on loaded dataset for particular setting of hyperparameters.
+        :param point: dict, pairs of parameter names and corresponding values
+        :return: score (more is better!) for some specific point
+        """
+
+        X, Y, Xv, Yv = self.dataset
```
Comment (betatim) [inline benchmarks/bench_ml.py]: is this the train test split? Maybe rename the variables `X_test` and `X_train`. If they are something else maybe add a comment to explain what they are :)

**Comment by betatim** [comment]
This looks good to me. Haven't tried to run it yet locally but I think this can be merged subject to some of my nitpicks over formatting. I will add those if there aren't any other comments to change the substance of the PR.

Maybe we can change the commit message to something more informative as well (if you want help with the right git magic spell to do that let me know).

(my nitpicks are basically about `Uppercase` variable names that aren't classes ;) )

File: benchmarks/bench_ml.py (hunk 1 above)
Comment (iaroslav-ai) [inline benchmarks/bench_ml.py]: fixed and commented

**Comment by iaroslav-ai** [comment]
Sounds good :) Will try to change the commit message, indeed not very informative.

**Comment by MechCoder** [comment]
the consensus seems to merge and then cleanup in another PR. Can you add the results of your benhmarks?

File: benchmarks/bench_ml.py [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+
+from skopt.learning import GaussianProcessRegressor, RandomForestRegressor, ExtraTreesRegressor
+from skopt.learning.gaussian_process.kernels import Matern
+from skopt import Optimizer
+from skopt.space import Real, Categorical, Integer
+
+import numpy as np
+import json
+from datetime import datetime
+from joblib import Parallel, delayed
```
Comment (betatim) [inline benchmarks/bench_ml.py]: can you import this from `sklearn.externals` instead

**Comment by MechCoder** [comment]
@iaroslav-ai Thanks for your work! It would be nice to abstract away the (dataset, model) creation and the black-box function creation. The static variables and the support method can be moved out of the class and just the initialization and the evaluate method can be kept inside the class.

**Comment by MechCoder** [comment]
Umm. I think this is long due for a merge. Merging for now and I'll clean up stuff in another PR.

**Comment by MechCoder** [comment]
Thanks!!

**Comment by glouppe** [comment]
Thanks @iaroslav-ai for your efforts!

**Comment by iaroslav-ai** [comment]
Thanks everyone for you comments and reviewing!
Also sorry for the delays lately - many things were happening.

**Comment by iaroslav-ai** [comment]
I will also run the experiments with all optimizers and PR the results.

**Pull request merged**